File: exercise/0-threads.c
#include <pthread.h>
#include <stdio.h>
#include <unistd.h>

const int NUM_THREADS = 5;

void *worker(void *arg) {
  int thread_id = *(int *)arg;

  printf("Thread %d is reading file...\n", thread_id);
  sleep(2);

  FILE *fp = fopen("0-threads-c.txt", "a");
  if (!fp) {
    perror("fopen");
    return NULL;
  }

  fprintf(fp, "Thread %d wrote this message.\n", thread_id);
  fclose(fp);

  return NULL;
}

int main(void) {
  pthread_t threads[NUM_THREADS];
  int thread_ids[NUM_THREADS];

  for (int i = 0; i < NUM_THREADS; i++) {
    thread_ids[i] = i;
    pthread_create(&threads[i], NULL, worker, (void *)&thread_ids[i]);
  }

  puts("Main thread doing some work asynchronously...");

  for (int i = 0; i < NUM_THREADS; i++)
    pthread_join(threads[i], NULL);

  puts("All threads have finished their work.");
}

File: exercise/0-threads.rs
use std::io::Write;

const NUM_THREADS: usize = 5;

fn worker(thread_id: usize) {
    println!("Thread {thread_id} is reading file...");
    std::thread::sleep(std::time::Duration::from_secs(2));

    std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open("0-threads-rs.txt")
        .expect("Failed to open file")
        .write_fmt(format_args!("Thread {thread_id} wrote this message\n"))
        .expect("Failed to write to file");
}

fn main() {
    let mut threads = Vec::new();

    for i in 0..NUM_THREADS {
        threads.push(std::thread::spawn(move || worker(i)));
    }

    println!("Main thread doing some work asynchronously...");

    for thread in threads {
        thread.join().expect("Thread panicked");
    }

    println!("All threads have finished their work.");
}
